nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -0,0 +1,169 @@
+import com.sun.tools.javac.code.Attribute;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.tree.JCTree;
+import java.util.HashSet;
+import java.util.List;
+
+/** Methods for performing checks related to generic types and nullability. */
+public class GenericsChecks {
+
+  private GenericsChecks() {
+    // just utility methods
+  }
+
+  /**
+   * Checks that for an instantiated generic type, {@code @Nullable} types are only used for type
+   * variables that have a {@code @Nullable} upper bound.
+   *
+   * @param type the instantiated type
+   * @param state the visitor state
+   * @param tree the tree in the AST representing the instantiated type
+   * @param analysis the analysis object
+   * @param config the analysis configuration
+   */
+  public static void checkInstantiatedType(
+      Type type, VisitorState state, Tree tree, NullAway analysis, Config config) {
+    CodeAnnotationInfo codeAnnotationInfo = CodeAnnotationInfo.instance(state.context);
+    if (!config.isJSpecifyMode()) {
+      return;
+    }
+    if (codeAnnotationInfo.isSymbolUnannotated(type.tsym, config)) {
+      return;
+    }
+
+    com.sun.tools.javac.util.List<Type> typeArguments = type.getTypeArguments();
+    HashSet<Integer> nullableTypeArguments = new HashSet<Integer>();
+    for (int index = 0; index < typeArguments.size(); index++) {
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrors =
+          typeArguments.get(index).getAnnotationMirrors();
+      boolean hasNullableAnnotation =
+          Nullness.hasNullableAnnotation(annotationMirrors.stream(), config);
+      if (hasNullableAnnotation) {
+        nullableTypeArguments.add(index);
+      }
+    }
+
+    Type baseType = type.tsym.type;
+    com.sun.tools.javac.util.List<Type> baseTypeArguments = baseType.getTypeArguments();
+    checkNullableTypeArgsAgainstUpperBounds(
+        state, tree, analysis, config, nullableTypeArguments, baseTypeArguments);
+    // Generics check for nested type parameters
+    for (Type typeArgument : typeArguments) {
+      if (typeArgument.getTypeArguments().length() > 0) {
+        checkInstantiatedType(typeArgument, state, tree, analysis, config);
+      }
+    }
+  }
+
+  /**
+   * Checks if the type arguments with a {@code @Nullable} annotation in an instantiated type have a
+   * {@code @Nullable} upper bound in the declaration, and reports an error otherwise.
+   *
+   * @param state the visitor state
+   * @param tree TODO not sure we need this
+   * @param analysis the analysis object
+   * @param config the analysis config
+   * @param nullableTypeArguments indices of {@code Nullable} type arguments
+   * @param baseTypeArguments list of type variables (with bounds) in the declared type
+   */
+  private static void checkNullableTypeArgsAgainstUpperBounds(
+      VisitorState state,
+      Tree tree,
+      NullAway analysis,
+      Config config,
+      HashSet<Integer> nullableTypeArguments,
```

Comment:
Here, rather than passing in a `HashSet<Integer>`, I think we should have a `Map<Integer,Tree>`, where the key is the index of the type parameter and the value is the actual `Tree` for the parameter.  This way, when we find a parameter that uses `@Nullable` incorrectly, we can use the `Tree` in the value as the tree for the error message and give an exact error location.